src/integer_mod_q/mat_zq/tensor.rs

```diff
@@ -0,0 +1,296 @@
+    fn vector_matrix() {
+        let vector = MatZq::from_str(&format!("[[1],[-1]] mod {}", u128::MAX)).unwrap();
+        let mat_1 = MatZq::from_str(&format!(
+            "[[1, {}, 1],[0, {}, -1]] mod {}",
+            u64::MAX,
+            i64::MAX,
+            u128::MAX
+        ))
+        .unwrap();
+
+        let mat_2 = vector.tensor_product(&mat_1);
+        let mat_3 = mat_1.tensor_product(&vector);
+        let mat_2_safe = vector.tensor_product_safe(&mat_1).unwrap();
+        let mat_3_safe = mat_1.tensor_product_safe(&vector).unwrap();
+
+        let cmp_mat_2 = MatZq::from_str(&format!(
+            "[[1, {}, 1],[0, {}, -1],[-1, -{}, -1],[0, -{}, 1]] mod {}",
+            u64::MAX,
+            i64::MAX,
+            u64::MAX,
+            i64::MAX,
+            u128::MAX
+        ))
+        .unwrap();
+        let cmp_mat_3 = MatZq::from_str(&format!(
+            "[[1, {}, 1],[-1, -{}, -1],[0, {}, -1],[0, -{}, 1]] mod {}",
+            u64::MAX,
+            u64::MAX,
+            i64::MAX,
+            i64::MAX,
+            u128::MAX
+        ))
+        .unwrap();
+
+        assert_eq!(cmp_mat_2, mat_2);
+        assert_eq!(cmp_mat_3, mat_3);
+        assert_eq!(cmp_mat_2, mat_2_safe);
+        assert_eq!(cmp_mat_3, mat_3_safe);
+    }
+
+    /// ensure that the tensor product works correctly with two vectors
+    #[test]
+    fn vector_vector() {
+        let vec_1 = MatZq::from_str(&format!("[[2],[1]] mod {}", u128::MAX)).unwrap();
+        let vec_2 = MatZq::from_str(&format!(
+            "[[{}],[{}]] mod {}",
+            (u64::MAX - 1) / 2,
+            i64::MIN / 2,
+            u128::MAX
+        ))
+        .unwrap();
+
+        let vec_3 = vec_1.tensor_product(&vec_2);
+        let vec_4 = vec_2.tensor_product(&vec_1);
+        let vec_3_safe = vec_1.tensor_product_safe(&vec_2).unwrap();
+        let vec_4_safe = vec_2.tensor_product_safe(&vec_1).unwrap();
+
+        let cmp_vec_3 = MatZq::from_str(&format!(
+            "[[{}],[{}],[{}],[{}]] mod {}",
+            u64::MAX - 1,
+            i64::MIN,
+            (u64::MAX - 1) / 2,
+            i64::MIN / 2,
+            u128::MAX
+        ))
+        .unwrap();
+        let cmp_vec_4 = MatZq::from_str(&format!(
+            "[[{}],[{}],[{}],[{}]] mod {}",
+            u64::MAX - 1,
+            (u64::MAX - 1) / 2,
+            i64::MIN,
+            i64::MIN / 2,
+            u128::MAX
+        ))
+        .unwrap();
+
+        assert_eq!(cmp_vec_3, vec_3);
+        assert_eq!(cmp_vec_4, vec_4);
+        assert_eq!(cmp_vec_3, vec_3_safe);
+        assert_eq!(cmp_vec_4, vec_4_safe);
+    }
+
+    /// ensure that entries are reduced by the modulus
+    #[test]
+    fn entries_reduced() {
+        let mat_1 = MatZq::from_str(&format!("[[1, 2],[3, 4]] mod {}", u64::MAX - 58)).unwrap();
+        let mat_2 = MatZq::from_str(&format!(
+            "[[1, {}],[0, {}]] mod {}",
+            u64::MAX,
+            u64::MAX - 59,
+            u64::MAX - 58
+        ))
```

Comment:
Optional: This makes it easier to understand in my opinion.
Something similar could also be done in the other string by using `-1`, `-2` ,`-3` , `-4` instead of `{}`.
```suggestion
let mat_2 = MatZq::from_str(&format!("[[1, 58],[0, -1]] mod {}", u64::MAX - 58)).unwrap();
```